File: Javascript/src/S1_SOLID/single-responsibility.ts
class Journal {
  entries: string[];
  count: number;

  constructor() {
    this.count = 0;
    this.entries = [];
  }

  addEntry(text: string) {
    this.entries.push(text);
    this.count++;
    return this.count;
  }

  removeEntry(index: number) {
    this.entries.splice(index, 1);
  }
}

class Persistence {
  saveToFile(journal: Journal, fileName: string) {
    console.log(journal.count);
    journal.entries.forEach((j) => console.log(j));
    console.log(`Saved into ${fileName} file.`);
  }
}

const journal = new Journal();
journal.addEntry('I cried today.');
journal.addEntry('I ate a bug.');

const p = new Persistence();
p.saveToFile(journal, 'Journal_1');

File: CSharp/S1_SOLID/SingleResponsibility.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace S1_SOLID
{
    public static class SingleResponsibility
    {
        public static void Main()
        {
            var j = new Journal();
            j.AddEntry("I cried today.");
            j.AddEntry("I ate a bug.");
            Console.WriteLine(j);

            var p = new Persistence();
            var filename = @"c:\temp\journal.txt";
            p.SaveToFile(j, filename);
            Process.Start(filename);
        }
    }

    public class Journal
    {
        private readonly List<string> entries = new List<string>();
        private static int count = 0;

        public int AddEntry(string text)
        {
            entries.Add($"{count++} {text}");
            return count;
        }

        public void RemoveEntry(int index)
        {
            entries.RemoveAt(index);
        }

        public override string ToString()
        {
            return string.Join(Environment.NewLine, entries);
        }

        // Three methods below broke Single Responsibility Principle
        public void Save(string filename, bool overwrite = false)
        {
            File.WriteAllText(filename, ToString());
        }
        public void Load(string filename)
        {

        }
        public void Load(Uri uri)
        {

        }
    }

    // Create a new class to handle the responsibility of persisting objects
    public class Persistence
    {
        public void SaveToFile(Journal journal, string filename, bool overwrite = false)
        {
            if (overwrite || !File.Exists(filename))
            {
                File.WriteAllText(filename, journal.ToString());
            }
        }
    }
}
